raft: move group0_state_machine::merger to its own header and add unit test for it

margdoc opened the pull request:
Move `merger` to its own header file. Leave the logic of applying commands to `group0_state_machine`. Remove `group0_state_machine` dependencies from `merger` to make it an independent module.

Add a test that checks if `group0_state_machine_merger` preserves timeuuid monotonicity. `last_id()` should be equal to the largest timeuuid, based on its timestamps.

This test combines two commands in the reverse order of their timeuuids. The timeuuids yield different results when compared in both timeuuid order and uuid order. Consequently, the resulting command should have a more recent timeuuid.

Closes #14568

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2465/

kbr-scylla commented on service/raft/group0_state_machine.cc:
```diff
@@ -77,169 +74,53 @@ static future<> write_mutations_to_database(storage_proxy& proxy, gms::inet_addr
-        }
+future<> group0_state_machine::merge_and_apply(group0_state_machine_merger& merger) {
+    auto [cmd, history] = merger.merge();
+    co_await apply_command(std::move(cmd), std::move(history));
```
subjective nit: I would inline `apply_command` here, since it's the only use of it.

kbr-scylla commented on test/boost/group0_cmd_merge_test.cc:
```diff
@@ -12,11 +12,53 @@
+}
+
+SEASTAR_TEST_CASE(test_group0_state_machine_merger_timeuuid_order) {
+    // https://github.com/scylladb/scylladb/issues/14568
```
We usually just reference the issue number with `#`, like this:
```
Ref #14568
```

and below:
```
From #7729
```

kbr-scylla replied:
Also, a better issue (or issues -- there are two in this case) to reference is those that described the bugs which were then solved, and this test is a regression test for the two issues. (The comment should explicitly say that this is a regression test for #X and #Y)

margdoc commented:
v2:
* inlined `apply_command`
* added comments

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2484/

margdoc commented:
v3:
* fixed documentation
* added `static` and `const` comments
* made `group0_state_machine_merger` a `class` instead of `struct`
* added information about merger refactoring in the commit message
* rebased on `master`

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2530/

kbr-scylla commented on service/raft/group0_state_machine_merger.hh:
```diff
@@ -0,0 +1,77 @@
+
+    // Returns true if the command can be merged with the current batch.
+    // Command can be merged if it is of the same type as commands in the current batch
+    // and the size of the batch will not exceed the limit. Empties the current batch.
```
I don't think `can_merge` empties the current batch.

kbr-scylla commented:
@kostja @tgrabiec please review

kbr-scylla commented:
Queued.
There was a minor problem with one comment, I fixed it during merge.
